src/validation.cpp

```diff
@@ -1394,6 +1394,11 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptPackage(const Package& package,
             // Provide the wtxid of the mempool tx so that the caller can look it up in the mempool.
             results_final.emplace(wtxid, MempoolAcceptResult::MempoolTxDifferentWitness(iter.value()->GetTx().GetWitnessHash()));
         } else {
+            if (wtxid == child->GetWitnessHash() && !quit_early) {
+                txns_package_eval.push_back(tx);
```

Comment:
> After this change it's impossible to hit quit_early=false and txns_package_eval.empty(), which was possible before.

Right, `quit_early=false && txns_package_eval.empty()` means everything so far has been valid or txid already in mempool.

> It means that even if everything was either (valid, already in the mempool, or invalid due to a subset of failures) — you will force the checks again? (before this change, it could have terminated early)

Ok I will try to break down all the possibilities in this scenario. If we get to the child and `quit_early=false && txns_package_eval.empty()`, everything else has been valid, and we should validate the child. It doesn't matter if we do it inside or outside this loop, but it's better to only do it 1 time.
The child's result can be: valid, policy/missing inputs, or other failure. If txid already in mempool we would have already exited the loop.

Before the commit:
- If valid: the tx was validated 1 time, inside the loop.
- If policy/missing inputs: the tx was validated 2 times, both inside the loop and with `AcceptMultipleTransactions(txns_package_eval)` outside the loop.
- If other failure: the tx was validated 1 time, inside the loop.

Right after the loop, `quit_early || txns_package_eval.empty()` hits, but none of them terminated without validating the child. 

After the commit:
- If valid: the tx is validated 1 time, outside the loop.
- If policy/missing inputs: the tx is validated 1 time, outside the loop.
- If other failure: the tx is validated 1 time, outside the loop.

Right after the loop, `quit_early || txns_package_eval.empty()` does not hit. The tx is validated, and then the results returned. The function returns at a later line of code, but that doesn't mean more work was done.

Reply:
Thank you for this elaborate analysis!

So in `if valid` and `if other failure` cases, it's not just inside vs outside the loop, right?
Because "outside the loop" also means re-validating other transactions, not just the child (`AcceptMultipleTransactions` call).
(Note that I'm look only at the first commit so far)

Reply:
What I'm talking about seems to be fixed in the next commit 879b55fa9da3688fb6b85e7f40f9778753a9102f.
If my understanding is correct, I'd suggest mentioning this in the commit message.

Reply:
Thank you yes, it makes more sense to switch the commits around. Switched their order